code/WorkInProgress/AzrunStuff.dm

```diff
@@ -58,4 +58,352 @@
+				/*  2*25 / 75 = 0.66 -2  = -1.34 		 TERRIBAD */
+				/* Use above offset * 10 to put it in the -20 to 40 ballpark */
+				efficiency_shift = (2 * electrical_conductivity / thermal_conductivity) - 2 //center on zero
+				efficiency_shift = clamp(efficiency_shift*10, -20, 40) //scale shift by 10 which gets it in the ballpark!
+				src.teg.efficiency_controller = clamp(src.teg.efficiency_controller + efficiency_shift, 25, 75) //ensure nothing goes bonkers
+
+			on_revert()
+				src.teg.efficiency_controller = prev_efficiency
+				. = ..()
+
+/*
+			 		if(W.material)
+			var/datum/teg_transformation/matsci/semiconductor = new
+			semiconductor.material = W.material.mat_id
+
+			if(generator.active_form)
+				generator.active_form.on_revert()
+			generator.active_form = semiconductor
+			generator.active_form.on_transform(generator)
+		*/
+
+/*
+  0         1         	2         	3        	4
+Present 	Unscrewed  Connected 	Unconnected		Missing
+     (Screw)------------------(Snip)--------(Pry)             >>-->> REMOVAL
+     (Screw)-----------------(Snip)-------(COIL)---(Item)		<<--<< REPLACEMNT
+*/
+
+/datum/action/bar/icon/teg_semiconductor_removal
+	id = "teg_semiconductor_removal"
+	interrupt_flags = INTERRUPT_MOVE | INTERRUPT_ACT | INTERRUPT_STUNNED | INTERRUPT_ACTION
+	duration = 200
+	icon = 'icons/ui/actions.dmi'
+	icon_state = "working"
+
+	var/obj/machinery/power/generatorTemp/generator
+	var/obj/item/the_tool
+
+	New(var/obj/O, var/obj/item/tool, var/duration_i)
+		..()
+		if (O)
+			generator = O
+		if (tool)
+			the_tool = tool
+			icon = the_tool.icon
+			icon_state = the_tool.icon_state
+		if (duration_i)
+			duration = duration_i
+		if (ishuman(owner))
+			var/mob/living/carbon/human/H = owner
+			if (H.traitHolder.hasTrait("training_engineer"))
+				duration = round(duration / 2)
+
+	onUpdate()
+		..()
+		if (generator == null || the_tool == null || owner == null || get_dist(owner, generator) > 1)
+			interrupt(INTERRUPT_ALWAYS)
+			return
+		var/mob/source = owner
+		if (istype(source) && the_tool != source.equipped())
+			interrupt(INTERRUPT_ALWAYS)
+
+	onStart()
+		..()
+		// SCREW->SNIP->CROW (REMOVAL)
+		if (generator.semiconductor_state == 0)
+			owner.visible_message("<span class='notice'>[owner] begins to dismantle \the [generator] to get access to the semiconductor.</span>")
+			playsound(get_turf(generator), "sound/items/Screwdriver.ogg", 50, 1)
+		if (generator.semiconductor_state == 1)
+			owner.visible_message("<span class='notice'>[owner] begins to snip wiring between the semiconductor and \the [generator].</span>")
+			playsound(get_turf(generator), "sound/items/Scissor.ogg", 60, 1)
+		if (generator.semiconductor_state == 3)
+			owner.visible_message("<span class='notice'>[owner] begins prying out the semiconductor from \the [generator].</span>")
+			playsound(get_turf(generator), "sound/items/Crowbar.ogg", 60, 1)
+
+	onEnd()
+		..()
+		// SCREW->SNIP->CROW (REMOVAL)
+		if (generator.semiconductor_state == 0)
```

Comment:
should use a switch

Reply:
Really great idea, makes if actually readable!